logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/Sequencer.java

```diff
@@ -0,0 +1,211 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.camunda.zeebe.logstreams.impl.log;
+
+import static io.camunda.zeebe.logstreams.impl.serializer.DataFrameDescriptor.FRAME_ALIGNMENT;
+
+import io.camunda.zeebe.logstreams.impl.serializer.DataFrameDescriptor;
+import io.camunda.zeebe.logstreams.log.LogAppendEntry;
+import io.camunda.zeebe.logstreams.log.LogStreamBatchWriter;
+import io.camunda.zeebe.scheduler.ActorCondition;
+import java.io.Closeable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.locks.ReentrantLock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The sequencer is a multiple-producer, single-consumer queue of {@link LogAppendEntry}. It buffers
+ * a fixed amount of entries and rejects writes when the queue is full. The consumer may read at its
+ * own pace by repeatedly calling {@link Sequencer#tryRead()} or register for notifications when new
+ * entries are written by calling {@link Sequencer#registerConsumer(ActorCondition)}.
+ *
+ * <p>The sequencer assigns all entries a position and makes that position available to its
+ * consumer. The sequencer does not copy or serialize entries, it only keeps a reference to them
+ * until they are handed off to the consumer.
+ */
+public final class Sequencer implements LogStreamBatchWriter, Closeable {
```

Comment:
:wrench: I think in general this class and most of its methods could have lowered visibility to package private. Lower/weaker visibility is useful to hint how your class should be used by consumers.